File: src/main/java/edu/allegheny/beecolony/Robot.java
import lejos.hardware.port.MotorPort;
import lejos.hardware.port.Port;
import lejos.hardware.port.SensorPort;
import lejos.hardware.motor.EV3LargeRegulatedMotor;
import lejos.hardware.sensor.EV3ColorSensor;
import lejos.robotics.SampleProvider;
import lejos.robotics.navigation.DifferentialPilot;

/**
 * Stores shared robot configuration.
 *
 * @author Hawk Weisman
 */
public class Robot {
    // -------------------------------------------------------------------------
    // PORT CONFIGURATION
    // -------------------------------------------------------------------------
    private static final Port SENSOR_COLOR = SensorPort.S1; // Change as appropriate
    private static final Port MOTOR_LEFT = MotorPort.B;
    private static final Port MOTOR_RIGHT = MotorPort.C;

    // -------------------------------------------------------------------------
    // WHEEL CONFIGURATION
    // -------------------------------------------------------------------------
    // These constants are based on some quick measurements from Lab 2 performed
    // using a plastic protractor with a precision of 1/16th inch. If the robot
    // doesn't turn accurately, please feel free to replace these w/ more
    // accurate measurements.
    private static final float WHEEL_DIAM = 2.0625f; // wheel diam ~= 2 1/6 in
    private static final float TRACK_WIDTH = 4.75f; // track width ~= 4 3/4 in

    // -------------------------------------------------------------------------
    // STATIC HARDWARE CLASSES
    // -------------------------------------------------------------------------
    public static final EV3ColorSensor colorSensor = new EV3ColorSensor(SENSOR_COLOR);
    private static final EV3LargeRegulatedMotor leftMotor = new EV3LargeRegulatedMotor(MOTOR_LEFT);
    private static final EV3LargeRegulatedMotor rightMotor = new EV3LargeRegulatedMotor(MOTOR_RIGHT);

    /**
     * Factory method for the color sensor color ID mode
     *
     * @return the color ID mode sensor provider
     */
    public static SampleProvider colorIDProvider() {
        return colorSensor.getColorIDMode();
    }

    /**
     * Factory method for a differential pilot with the robot's current
     * configuration.
     *
     * @return a DifferentialPilot configured for the robot's current
     *         configuration.
     */
    public static DifferentialPilot pilot() {
        return new DifferentialPilot(WHEEL_DIAM, TRACK_WIDTH, leftMotor,
                rightMotor);
    }

}

File: src/main/scala/edu/allegheny/beecolony/Robot.scala
package edu.allegheny.beecolony

import java.net.Socket

import lejos.hardware.motor.EV3LargeRegulatedMotor
import lejos.hardware.port.{MotorPort, SensorPort}
import lejos.hardware.sensor.EV3ColorSensor
import lejos.robotics.navigation.{DifferentialPilot,Navigator}

import scala.annotation.tailrec
import scala.language.postfixOps
import scala.util.{Try,Success,Failure}

/**
 * Stores shared robot configuration.
 *
 * @author Hawk Weisman
 */
trait Robot {

  type Coordinate = Tuple2[Float,Float]
  type Color = Int

  // -------------------------------------------------------------------------
  // PORT CONFIGURATION
  // -------------------------------------------------------------------------
  private val SENSOR_COLOR  = SensorPort.S1; // Change as appropriate
  private val MOTOR_LEFT    = MotorPort.B
  private val MOTOR_RIGHT   = MotorPort.C

  // -------------------------------------------------------------------------
  // WHEEL CONFIGURATION
  // -------------------------------------------------------------------------
  // These constants are based on some quick measurements from Lab 2 performed
  // using a plastic protractor with a precision of 1/16th inch. If the robot
  // doesn't turn accurately, please feel free to replace these w/ more
  // accurate measurements.
  private val WHEEL_DIAM  = 2.0625f
  private val TRACK_WIDTH = 4.75f

  // -------------------------------------------------------------------------
  // LeJOS CLASSES
  // -------------------------------------------------------------------------
  private val colorSensor = new EV3ColorSensor(SENSOR_COLOR)
  private val leftMotor   = new EV3LargeRegulatedMotor(MOTOR_LEFT)
  private val rightMotor  = new EV3LargeRegulatedMotor(MOTOR_RIGHT)
  private val colorIDProvider = colorSensor.getColorIDMode
  private val pilot           = new DifferentialPilot(WHEEL_DIAM, TRACK_WIDTH, leftMotor, rightMotor)
  private val nav             = new Navigator(pilot)
  private val poseProvider    = nav getPoseProvider

  private val sample = new Array[Float](1) // this is because the LeJOS api is awful
  /**
   * Move the robot to a coordinate pair
   * @param where the coordinate to move to
   */
  def goTo(where: Coordinate): Unit = { (x: Float,y: Float) => nav goTo (x,y) }

  /**
   * @return the current coordinates
   */
  def location: Coordinate  = (poseProvider.getPose getX, poseProvider.getPose getY)

  /**
   * @return the color ID currently under the color sensor
   */
  def checkColor: Color = {
    colorIDProvider fetchSample (sample, 0) // ugh, I had to write an impure function
    sample(0) toInt
  }

}

